src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,335 @@
+  ∃ (a0 b0 c0), (minimal a0 b0 c0) ∧ a0 % 2 = 1 :=
+begin
+  obtain ⟨a0, b0, c0, hf⟩ := exists_minimal h,
+  cases int.mod_two_eq_zero_or_one a0 with hap hap,
+  { cases int.mod_two_eq_zero_or_one b0 with hbp hbp,
+    { exfalso,
+      have h1 : 2 ∣ (int.gcd a0 b0 : ℤ),
+      { exact int.dvd_gcd (int.dvd_of_mod_eq_zero hap) (int.dvd_of_mod_eq_zero hbp) },
+        rw (coprime_of_minimal hf) at h1, revert h1, norm_num },
+      { exact ⟨b0, ⟨a0, ⟨c0, minimal_comm hf, hbp⟩⟩⟩ }},
+  exact ⟨a0, ⟨b0, ⟨c0 , hf, hap⟩⟩⟩,
+end
+
+/-- We can assume that a minimal solution to `a ^ 4 + b ^ 4 = c ^ 2` has
+`a` odd and `c` positive. -/
+lemma exists_pos_odd_minimal {a b c : ℤ} (h : fermat_42 a b c) :
+  ∃ (a0 b0 c0), (minimal a0 b0 c0) ∧ a0 % 2 = 1 ∧ 0 < c0  :=
+begin
+  obtain ⟨a0, b0, c0, hf, hc⟩ := exists_odd_minimal h,
+  by_cases h0 : 0 < c0,
+  { use [a0, b0, c0], tauto },
+  { use [a0, b0, -c0],
+    apply and.intro (neg_of_minimal hf),
+    apply and.intro hc,
+    apply neg_pos.mpr, apply not_le.mp, intro h2,
+    have h3 : 0 < c0, { exact lt_of_le_of_ne h2 (ne.symm (ne_zero hf.1)) },
+    revert h3, exact h0 }
+end
+
+end fermat_42
+
+lemma int.coprime_of_sqr_sum {m r s : ℤ} (h : m = r ^ 2 + s ^ 2) (h2 : int.gcd r s = 1) :
+  int.gcd m r = 1 :=
+begin
+  by_contradiction hg,
+  obtain ⟨p, ⟨hp, hm, hr⟩⟩ := nat.prime.not_coprime_iff_dvd.mp hg,
+  apply nat.prime.not_dvd_one hp,
+  rw ← h2,
+  apply nat.dvd_gcd hr,
+  apply int.coe_nat_dvd_left.mp,
+  apply (or_self _).mp (int.prime.dvd_mul' hp _),
+  rw [← pow_two, (eq_sub_of_add_eq' (eq.symm h))],
+  apply dvd_sub (int.coe_nat_dvd_left.mpr hm),
+  rw pow_two, exact dvd_mul_of_dvd_left (int.coe_nat_dvd_left.mpr hr) r
+end
+
+lemma int.coprime_of_sqr_sum' {m r s : ℤ} (h : m = r ^ 2 + s ^ 2) (h2 : int.gcd r s = 1) :
+  int.gcd m (r * s) = 1 :=
+begin
+  dunfold int.gcd,
+  rw int.nat_abs_mul,
+  apply nat.coprime.mul_right (int.coprime_of_sqr_sum h h2),
+  rw add_comm at h,
+  rw [int.gcd_comm r s] at h2,
+  exact int.coprime_of_sqr_sum h h2
+end
+
+namespace fermat_42
+
+-- If we have a solution to a ^ 4 + b ^ 4 = c ^ 2, we can construct a smaller one. This
+-- implies there can't be a smallest solution.
+lemma not_minimal {a b c : ℤ}
+  (h : minimal a b c) (ha2 : a % 2 = 1) (hc : 0 < c) : false :=
+begin
+  -- Use the fact that a ^ 2, b ^ 2, c form a pythagorean triple to obtain m and n such that
+  -- a ^ 2 = m ^ 2 - n ^ 2, b ^ 2 = 2 * m * n and c = m ^ 2 + n ^ 2
+  -- first the formula:
+  have ht : pythagorean_triple (a ^ 2) (b ^ 2) c,
+  { calc ((a ^ 2) * (a ^ 2)) + ((b ^ 2) * (b ^ 2))
+        = a ^ 4 + b ^ 4 : by ring
+    ... = c ^ 2 : by rw h.1.2.2
+    ... = c * c : by rw pow_two
+  },
+  -- coprime requirement:
+  have h2 : int.gcd (a ^ 2) (b ^ 2) = 1,
+  { change nat.coprime (int.nat_abs (a ^ 2)) (int.nat_abs (b ^ 2)),
+    rw [pow_two, int.nat_abs_mul],
+    apply nat.coprime.mul;
+    { rw [pow_two, int.nat_abs_mul], apply nat.coprime.mul_right;
```

Comment:
Why two tactics on one line?

Reply:
Is that a no-no? I wasn't aware of this... will look out for it in future.